cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,356 @@
+  double const* min_val = min_ + tdigest_index;
+  double const* max_val = max_ + tdigest_index;
+
+  double const total_weight = cumulative_weight[tdigest_size - 1];
+  double const cluster_q    = percentage * total_weight;
+
+  cast_output<T> cast_op{output_type};
+
+  if (cluster_q <= 1) {
+    output[tid] = cast_op(*min_val);
+    return;
+  } else if (cluster_q >= total_weight - 1) {
+    output[tid] = cast_op(*max_val);
+    return;
+  }
+
+  // otherwise find the centroid we're in and interpolate
+  size_type const centroid_index = static_cast<size_type>(
+    thrust::lower_bound(
+      thrust::seq, cumulative_weight, cumulative_weight + tdigest_size, cluster_q) -
+    cumulative_weight);
+
+  double diff = cluster_q + weight[centroid_index] / 2 - cumulative_weight[centroid_index];
+  if (weight[centroid_index] == 1 && std::abs(diff) < 0.5) {
+    output[tid] = cast_op(mean[centroid_index]);
+    return;
+  }
+  size_type left_index  = centroid_index;
+  size_type right_index = centroid_index;
+  if (diff > 0) {
+    if (right_index == tdigest_size - 1) {
+      output[tid] = cast_op(lerp(mean[right_index], *max_val, diff / (weight[right_index] / 2)));
+      return;
+    }
+    right_index++;
+  } else {
+    if (left_index == 0) {
+      output[tid] = cast_op(lerp(*min_val, mean[left_index], diff / (weight[left_index] / 2)));
+      return;
+    }
+    left_index--;
+    diff += weight[left_index] / 2 + weight[right_index] / 2;
+  }
+
+  diff /= (weight[left_index] / 2 + weight[right_index] / 2);
+  output[tid] = cast_op(lerp(mean[left_index], mean[right_index], diff));
+}
+
+/**
+ * @brief Calculate approximate percentiles on a provided tdigest column.
+ *
+ * Produces a LIST column where each row N represents output from querying the
+ * corresponding tdigest of from row N in `input`. The length of each output list
+ * is the number of percentiles specified in `percentiles`
+ *
+ * @param input           tdigest input data. One tdigest per row.
+ * @param percentiles     Desired percentiles in range [0, 1].
+ * @param stream          CUDA stream used for device memory operations and kernel launches
+ * @param mr              Device memory resource used to allocate the returned column's device
+ * memory
+ *
+ * @returns Column of doubles containing requested percentile values.
+ */
+template <typename T>
+std::unique_ptr<column> compute_approx_percentiles(structs_column_view const& input,
+                                                   column_view const& percentiles,
+                                                   data_type output_type,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view lcv(input.child(tdigest_centroid_column_index));
+  column_view min_col = input.child(tdigest_min_column_index);
+  column_view max_col = input.child(tdigest_max_column_index);
```

Comment:
The `tdigests_column_view` could simplify this stuff too.

Reply:
Oh, I see. I thought we weren't going with this, given that it's used only here.

Reply:
well it's used in at least two places: creating the tdigest aggregation and computing the `percentile_approx`.

Even if it's only ever used in those two places, I believe creating a `tdigests_column_view` type is worthwhile as it is very low effort and encapsulates information that is otherwise defined in multiple places.

Reply:
As above, suggest doing this as a followup PR.